acp/acp_local.go

```diff
@@ -0,0 +1,328 @@
+	}
+
+	l.localModule = &localACPModule
+	l.pathToStore = acpStorePath
+
+	return nil
+}
+
+func (l *ACPLocal) Close() error {
+	return l.localModule.Close()
+}
+
+func (l *ACPLocal) AddPolicy(
+	ctx context.Context,
+	creatorID string,
+	policy string,
+) (string, error) {
+	// Having a creator identity is a MUST requirement for adding a policy.
+	if creatorID == "" {
+		return "", ErrPolicyCreatorMustNotBeEmpty
+	}
+
+	if policy == "" {
+		return "", ErrPolicyDataMustNotBeEmpty
+	}
+
+	var createPolicy types.MsgCreatePolicy
+
+	// Assume policy is in YAML format by default.
+	policyMarshalType := types.PolicyMarshalingType_SHORT_YAML
+	if isJSON := fastjson.Validate(policy) == nil; isJSON { // Detect JSON format.
+		policyMarshalType = types.PolicyMarshalingType_SHORT_JSON
+	}
+
+	createPolicy = types.MsgCreatePolicy{
+		Creator:      creatorID,
+		Policy:       policy,
+		MarshalType:  policyMarshalType,
+		CreationTime: protoTypes.TimestampNow(),
+	}
+
+	createPolicyResponse, err := l.localModule.GetMsgService().CreatePolicy(
+		l.localModule.GetCtx(),
+		&createPolicy,
+	)
+
+	if err != nil {
+		return "", NewErrFailedToAddPolicyWithACP(err, "Local", creatorID)
+	}
+
+	policyID := createPolicyResponse.Policy.Id
+	log.InfoContext(ctx, "Created Policy", corelog.Any("PolicyID", policyID))
+
+	return policyID, nil
+}
+
+func (l *ACPLocal) ValidateResourceExistsOnValidDPI(
+	ctx context.Context,
+	policyID string,
+	resourceName string,
+) error {
+	if policyID == "" && resourceName == "" {
+		return ErrNoPolicyArgs
+	}
+
+	if policyID == "" {
+		return ErrPolicyIDMustNotBeEmpty
+	}
+
+	if resourceName == "" {
+		return ErrResourceNameMustNotBeEmpty
+	}
+
+	queryPolicyRequest := types.QueryPolicyRequest{Id: policyID}
+	queryPolicyResponse, err := l.localModule.GetQueryService().Policy(
+		l.localModule.GetCtx(),
+		&queryPolicyRequest,
+	)
+
+	if err != nil {
+		if errors.Is(err, types.ErrPolicyNotFound) {
+			return newErrPolicyIDValidation(
+				err,
+				policyID,
+				errPolicyDoesNotExistOnACPModule,
+			)
+		} else {
+			return newErrPolicyIDValidation(
+				err,
+				policyID,
+				errPolicyValidationFailedOnACPModule,
+			)
+		}
+	}
+
+	// So far we validated that the policy exists, now lets validate that resource exists.
+	resourceResponse := queryPolicyResponse.Policy.GetResourceByName(resourceName)
+	if resourceResponse == nil || resourceResponse.Name != resourceName {
+		return newErrResourceDoesNotExistOnTargetPolicy(err, resourceName, policyID)
+	}
+
+	// Now that we have validated that policyID exists and it contains a corresponding
+	// resource with the matching name, validate that all required permissions
+	// for DPI actually exist on the target resource.
+	for _, requiredPermission := range dpiRequiredPermissions {
+		permissionResponse := resourceResponse.GetPermissionByName(requiredPermission)
+		if permissionResponse == nil || permissionResponse.Name != requiredPermission {
```

Comment:
> question: Why would `GetPermissionByName(name)` return a policy of a different name than the one provided to it?
> 
> suggestion: Remove `|| permissionResponse.Name != requiredPermission`.
> 
> Same question/suggestion for the `GetResourceByName` call above too.

HAHA because I have trust issues. I can remove the extra check.

- [ ] Fix trust issues